### PR TITLE
Align KTO with DPO: Move completion assembly from _prepare_dataset to data collator

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Move completion assembly from `_prepare_dataset` to `DataCollatorForUnpairedPreference`.

This PR refactors the KTOTrainer data pipeline to simplify and standardize how prompts and completions are represented and processed. It moves completion assembly out of `_prepare_dataset` and into `DataCollatorForUnpairedPreference`, aligning the KTO data pipeline with DPO's: the dataset stores raw token IDs, and the collator handles truncation, BOS/EOS insertion, sequence assembly, and label creation at batch time. It replaces the previous use of multiple keys (such as `*_input_ids`, `*_attention_mask`, and `*_labels`) with a cleaner, more explicit use of `prompt_ids` and `completion_ids`. The data collator is also rewritten to assemble and pad sequences, add BOS/EOS tokens, and generate labels in a more modular way. The test suite and KL dataset handling are updated accordingly.

Part of:
- #4786

### Changes

**Refactoring of data representation and processing:**

* The data pipeline now consistently uses `prompt_ids` and `completion_ids` instead of multiple input/attention/label keys. This simplifies data handling and makes the dataset format clearer.
* The data collator (`DataCollatorForUnpairedPreference`) gains `max_length`, `bos_token_id`, and `eos_token_id` fields. The `torch_call` is rewritten to assemble completions from raw token IDs, handle BOS/EOS tokens, truncate sequences as needed, and generate labels with the correct masking.
* KL dataset rotation and assembly are simplified: the KL dataset is now created by rotating `completion_ids` only, and a simple renaming function is used instead of re-assembling sequences. 
* The `tokenize_fn`** is reduced to two output fields: `prompt_ids` (raw tokenized prompt) and `completion_ids` (raw answer tokens, no prompt prefix), matching DPO's `tokenize_fn` output style.
* The intermediate `answer_input_ids` / `answer_attention_mask` columns, and the corresponding `remove_columns` call, are gone: those fields are never created.

**Test and documentation updates:**

* The test suite is updated to match the new data format, checking that `prompt_ids` and `completion_ids` are correctly populated and that the collator output is as expected.
* Documentation and logging messages are updated to reference the new data collator and column names, ensuring clarity for users.

These changes make the KTOTrainer codebase more maintainable and less error-prone by unifying the handling of prompt and completion data throughout the pipeline.

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it refactors KTO’s input pipeline (tokenization, truncation, BOS/EOS handling, KL rotation), which can subtly change training/eval behavior and loss masking if any edge cases are missed.
> 
> **Overview**
> Refactors KTO data preparation to **store only raw `prompt_ids` and `completion_ids` in the dataset** and moves sequence assembly (prompt+completion), truncation, BOS/EOS insertion, attention masks, and label masking into `DataCollatorForUnpairedPreference` at batch time.
> 
> Simplifies KL dataset handling by rotating `completion_ids` only and carrying them as `KL_completion_ids` for the collator to assemble, removing the previous intermediate `answer_*` and pre-assembled `*_input_ids/*_labels` columns. Tests are updated to validate the new dataset format and to assert collator-produced `completion_*` tensors (including EOS and `-100` prompt masking).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6b21e788237af7d4c6685c6b8c90888dd084b895. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — COMMENTED** [review]
super important one, thanks!

**Comment by qgallouedec** [comment]
@codex review

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -89,46 +89,102 @@ def _get_kl_dataset(batch: dict[str, list[Any]]) -> dict[str, list[Any]]:
...
+                completion_ids = prompt_ids + answer_ids
+
+                # Add BOS, which affects both prompt and the full completion
+                if self.bos_token_id is not None and (not prompt_ids or prompt_ids[0] != self.bos_token_id):
+                    prompt_ids = [self.bos_token_id] + prompt_ids
+                    completion_ids = [self.bos_token_id] + completion_ids
+
+                # Add EOS, which affects only the full completion
+                if not answer_ids or self.eos_token_id != answer_ids[-1]:
+                    completion_ids = completion_ids + [self.eos_token_id]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/kto/kto_trainer.py]: Eos should always be set

File: trl/experimental/kto/kto_trainer.py [hunk 2]
```diff
@@ -89,46 +89,102 @@ def _get_kl_dataset(batch: dict[str, list[Any]]) -> dict[str, list[Any]]:
...
+            Token ID for the end-of-sequence token. If set and not already present at the end of `completion_ids`, it
+            is appended to the completion.
+        return_tensors (`str`, *optional*, defaults to `"pt"`):
             The tensor type to return. Currently, only `"pt"` (PyTorch tensors) is supported.
     """
 
     pad_token_id: int = 0
+    max_length: int | None = None
+    bos_token_id: int | None = None
+    eos_token_id: int | None = None
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py]: It's kinda strange to handle eos/bos in the collator, in my opinion, it's more a data preparation operation. I'm fine merging it, but I'd aim to have this in the `prepare_dataset` instead in the future

**Review by qgallouedec — APPROVED** [review]

File: trl/experimental/kto/kto_trainer.py (hunk 2 above)
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py]: I am planning to remove all that logic in my subsequent PR, so we are completely aligned with DPO.
